### DISCUSSION TITLE
Build Autofixture Customization Entirely from Reflection

### DISCUSSION THREAD

**Discussion by hughesjs** [AutoFixture/AutoFixture#1353] [Q&A]
I've [posted this question on Stack Overflow](https://stackoverflow.com/questions/72676767/build-autofixture-customization-using-reflection) but figured it's probably too niche to get an answer. So I've recreated it here in the hopes that someone can help.

------

I'm trying to create an `AutoFixture.ICustomization` that loops over every type that inherits from my model base and then explicitly doesn't assign any values to properties that also derive from that model base.

Obviously, this requires some serious abuse of reflection.

So far, I've got this:

```cs
public class DiscosModelFixtureCustomizationNoLinks: ICustomization
{
	public void Customize(IFixture fixture)
	{
		Type[]     discosModelTypes = typeof(DiscosModelBase).Assembly.GetTypes().Where(t => t.IsDiscosModel() && !t.IsAbstract).ToArray();
		MethodInfo customizeMethod  = fixture.GetType().GetMethods().Single(m => m.Name == nameof(fixture.Customize) && m.IsGenericMethod); // Needed to resolve overload with generic
		
		foreach (Type t in discosModelTypes)
		{
			MethodInfo     constructedCustomizeMethod = customizeMethod.MakeGenericMethod(t);
			Type           customizationComposer      = typeof(ICustomizationComposer<>).MakeGenericType(t);
			PropertyInfo[] propsToIgnore              = t.GetProperties().Where(p => p.PropertyType.IsDiscosModel()).ToArray();
			foreach (PropertyInfo prop in propsToIgnore)
			{
				// I want to essentially do this
                // For every prop that derives from DiscosModelBase
				// fixture.Customize<DiscosObject>(c => c.Without(p => p.Id));
				
				// Using my constructed method
                constructedCustomizeMethod.Invoke(fixture, new [] {myExpression});
			}
		}
	}
}
```

So `.Customize<T>()` has this signature:

```cs
void Customize<T>(Func<ICustomizationComposer<T>, ISpecimenBuilder> composerTransformation);
```

And `ICustomizationComposer<T>.Without()` has this signature:

```cs
IPostprocessComposer<T> Without<TProperty>(Expression<Func<T, TProperty>> propertyPicker);
```

So essentially, I need to work out how to build the equivalent of `c => c.Without(p => p.Id)` using reflection and my `PropertyInfo`.

[This answer][1] would indicated that I can manually build the expression tree, however, it's not really explained in enough detail that I'm able to apply it in this context.

First up, is this possible? Secondly, if so, how do I accomplish this missing step?

  [1]: https://stackoverflow.com/a/30916533/4700841

**Comment by aivascu** [comment]
Hello @hughesjs 
Can you provide an example of the test that you're trying to write? 
If I understood your question, you want to ignore a specific property on all types inheriting from a base type.
This sounds like you need to use an omitter with a property specification that matches reflected types that inherit your base type.

**Comment by hughesjs** [comment]
Hi mate, thanks for getting back to me!

This customization is going to be provided as part of an SDK I'm writing to facilitate all sorts of tests. The reason being, the API that the SDK is for conforms to the JSON API spec so, by default, it only returns the data of an object itself, not the data of other related objects. These relatives have to be explicitly included. Thus, for most queries, these will be empty/null. Take, this class for example that represents an orbital launch vehicle:

```cs
public record LaunchVehicle : DiscosModelBase
{
        // These should all be filled
	public int? FailedLaunches { get; init; }
	public int? SuccessfulLaunches { get; init; }
	public int? NumStages { get; init; }
	public float? GeoCapacity { get; init; }
	public float? EscCapacity { get; init; }
	public float? SsoCapacity { get; init; }
	public float? LeoCapacity { get; init; }
	public float? GtoCapacity { get; init; }
	public float? Height { get; init; }
	public float? Diameter { get; init; }
	public float? Mass { get; init; }
	public float? ThrustLevel { get; init; }

        // These should not
	public IReadOnlyList<LaunchVehicleStage> Stages { get; init; } = ArraySegment<LaunchVehicleStage>.Empty;
	public IReadOnlyList<Launch> Launches { get; init; } = ArraySegment<Launch>.Empty;
	public IReadOnlyList<LaunchVehicleEngine> Engines { get; init; } = ArraySegment<LaunchVehicleEngine>.Empty;
	public LaunchVehicleFamily? Family { get; init; }
}
```

If I query the API for a launch vehicle with no additional params, the Stages, Launches, Engines and Family won't be fetched but everything else will be. Everything that shouldn't be fetched, either derives from `DiscosModelBase` or is a collection of `DiscosModelBase`, I have this extension method that determines if a given type conforms to this:

```cs
public static class TypeExtensions
{
	public static bool IsDiscosModel(this Type t) =>
		t.IsAssignableTo(typeof(DiscosModelBase))                                           ||
		t.HasElementType     && t.GetElementType()!.IsAssignableTo(typeof(DiscosModelBase)) ||
		t.IsCollectionType() && t.IsGenericType && t.GetGenericArguments().Single().IsAssignableTo(typeof(DiscosModelBase));
}
```

So I've actually written a test for this customization that I think should do the trick. Although, it's obviously a contrived example and I haven't checked it works yet:

```cs
public class DiscosModelFixtureCustomizationNoLinksTests
{
	private readonly Fixture _fixture;

	public DiscosModelFixtureCustomizationNoLinksTests()
	{
		_fixture = new();
		_fixture.Customize(new DiscosModelFixtureCustomizationNoLinks());
	}
	
	[Theory]
	[ClassData(typeof(DiscosModelTypesTestDataGenerator))]
	public void CanCreateAllDiscosModels(Type t, string _)
	{
		DiscosModelBase model = (DiscosModelBase)new SpecimenContext(_fixture).Resolve(t);
		model.GetType().GetProperties().Select(p => p.GetValue(model)).ShouldNotBeNull();
	}
}
```

I will be writing a second customizer that does fill in these links, but only to the depth of a single object as 1) there are circular paths and 2) recursive fetching isn't supported by the API anyway but that's a future problem that I suspect will be easy enough once I know how to do this one!

I hope that makes things a little clearer?

> This sounds like you need to use an omitter with a property specification that matches reflected types that inherit your base type.

That last sentence sounds exactly like what I'm trying to do, I just can't work out quite how to do it, as you can see from my attempt above.

**Comment by aivascu** [comment]
@hughesjs the way to ignore a property in AutoFixture is to provide it a customization (`ISpecimenBuilder` implementation) that will instruct it to not set the value for this property, essentially an omitter customization. AutoFixture provides out of the box the `Omitter` type which does exactly that. The only thing it requires is a specification (`IRequestSpecification`) which will identify the request to be omitted. The specification requires only one method to be implemented `bool IsSatisfiedBy(object)`. The method returns true if the request matches the specification and false if it doesn't. There are a few specifications already provided by AutoFixture for reflection work.

I'm going to simplify a bit the example you provided, for brevity. Let's say you have two types, a `BaseType` which contains a couple properties which you always want ignored by AutoFixture and a `ChildType` of which you want just some properties to be filled in.

```cs
// Everything in this type should be omitted
public abstract record BaseType
{
    public decimal? Decimal { get; set; }
}

public record ChildType : BaseType
{
    public string StringValue { get; set; }
    public float? FloatValue { get; set; }

    // This property should be omitted
    public IReadOnlyList<int> Numbers { get; set; }
}
```
First let's write a specification that will ignore everything in `BaseType`. AutoFixture already provides a basic specification for matching properties `PropertySpecification` but it requires a `IEquatable<PropertyInfo>` instance which will tell if the property matches the specification. So let's write a comparer that matches all properties in a declaring type.

```cs
public class DeclaringTypeComparer : IEquatable<PropertyInfo>
{
    public DeclaringTypeComparer(Type declaringType)
    {
        this.DeclaringType = declaringType;
    }

    public Type DeclaringType { get; }

    public bool Equals(PropertyInfo other)
    {
        return other.DeclaringType == this.DeclaringType;
    }
}
```
Next our specification will look like this
```cs
var spec = new PropertySpecification(new DeclaringTypeComparer(typeof(BaseType)))
```
we can use it with the omitter type to create a customization that will ignore all properties inherited from `BaseType`
```cs
var omitter = new Omitter(new PropertySpecification(new DeclaringTypeComparer(typeof(BaseType))));
var fixture = new Fixture();
fixture.Customizations.Add(omitter);
```
Now to ignore particular properties on a specific type we can make use of the fact that specifications can be combined using And and Or clauses.
```cs
var spec = new OrRequestSpecification(
    new PropertySpecification(new DeclaringTypeComparer(typeof(BaseType))),
    new AndRequestSpecification(
        new PropertySpecification(
            new DeclaringTypeComparer(typeof(ChildType))),
        new PropertySpecification(
            new PropertyTypeAndNameCriterion(
                new Criterion<Type>(typeof(IReadOnlyList<int>), EqualityComparer<Type>.Default),
                new Criterion<string>("Numbers", StringComparer.Ordinal)))));
```
Now using the `Omitter` as before we can write this test that will ignore all properties inherited from base type and one specific property of a child type.
```cs
[Fact]
public void Foo()
{
    var spec = new OrRequestSpecification(
        new PropertySpecification(new DeclaringTypeComparer(typeof(BaseType))),
        new AndRequestSpecification(
            new PropertySpecification(
                new DeclaringTypeComparer(typeof(ChildType))),
            new PropertySpecification(
                new PropertyTypeAndNameCriterion(
                    new Criterion<Type>(typeof(IReadOnlyList<int>), EqualityComparer<Type>.Default),
                    new Criterion<string>("Numbers", StringComparer.Ordinal)))));
    var omitter = new Omitter(spec);
    var fixture = new Fixture();
    fixture.Customizations.Insert(0, omitter);

    var sut = fixture.Create<ChildType>();

    Assert.NotNull(sut.StringValue);
    Assert.NotNull(sut.FloatValue);
    Assert.Null(sut.Numbers);
    Assert.Null(sut.Decimal);
}
```

Let me know if this helps.

**Comment by hughesjs** [comment]
Huh, looks like I was massively overcomplicating things! Thanks my dude.

Managed to get it working with this:
```cs
public class DiscosModelFixtureCustomizationNoLinks: ICustomization
{
	public void Customize(IFixture fixture)
	{
		var omitter = new Omitter(new PropertySpecification(new DiscosModelPropertyComparer()));
		fixture.Customizations.Add(omitter);
	}
	
	private class DiscosModelPropertyComparer : IEquatable<PropertyInfo>
	{
		public bool Equals(PropertyInfo other) => other.PropertyType.IsDiscosModel();
	}
}
```

**Reply by aivascu** [reply]
@hughesjs glad to help :)

**Reply by hughesjs** [reply]
I'm going to write up this discussion as a SO answer for my question, I hope that's alright? I'll link it back to this discussion.
